pysb/bng.py

```diff
@@ -91,14 +92,260 @@ class GenerateNetworkError(RuntimeError):
     """BNG reported an error when trying to generate a network for a model."""
     pass
 
-_generate_network_code = """
-begin actions
-generate_network({overwrite=>1})
-end actions
-"""
 
+class BngBaseInterface(object):
+    """ Abstract base class for interfacing with BNG """
+    __metaclass__ = abc.ABCMeta
+
+    def __init__(self, model, verbose=False, cleanup=False,
+                 output_prefix=None):
+        self.verbose = verbose
+        self.cleanup = cleanup
+        self.output_prefix = 'tmpBNG' if output_prefix is None else \
+            output_prefix
+        self.generator = BngGenerator(model)
+        self._check_model()
+
+    def __enter__(self):
+        return self
 
-def run_ssa(model, t_end=10, n_steps=100, param_values=None, output_dir=os.getcwd(),
+    @abc.abstractmethod
+    def __exit__(self):
+        return
+
+    def _check_model(self):
+        """
+        Checks a model has at least one initial condition and rule, raising
+        an exception if not
+        """
+        if not self.model.initial_conditions:
+            raise NoInitialConditionsError()
+        if not self.model.rules:
+            raise NoRulesError()
+
+    @classmethod
+    def _bng_param(cls, param):
+        """
+        Ensures a BNG console parameter is in the correct format
+
+        Strings are double quoted and booleans are mapped to [0,1]. Other
+        types are currently used verbatim.
+
+        Parameters
+        ----------
+        param :
+            An argument to a BNG action call
+        """
+        if isinstance(param, str):
+            return '"%s"' % param
+        if isinstance(param, bool):
+            return 1 if param else 0
+        return param
+
+    @abc.abstractmethod
+    def action(self, action, **kwargs):
```

Comment:
This is a misuse of `abstractmethod` -- this `action` and its implementation in subclasses do different things. This one should have a different name, like `format_action`, and be a `classmethod`. Then `action` would be an empty `abstractmethod`. The subclasses would then call `self.format_action` to build the `action_args` instead of `super(...).action`.